keps/sig-network/2009-dns-autopath/README.md

```diff
@@ -0,0 +1,228 @@
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+DNS Search Path expansion on k8s is one of the top reasons for high DNS latency and query volume. This proposal aims to minimize the number of parallel DNS queries generated by a pod by moving searchpath expansion logic to the DNS server side. 
+
+## Motivation
+
+DNS Search Path expansion on pods using ClusterFirst DNS mode can lead to DNS latency issues and race conditions due to several parallel (The queries are sent in parallel in musl) dns queries from the same pod. This is especially problematic for non-cluster service lookups, where all expanded queries will be NXDOMAINs. Even in pods using glibc which sends these requests serially, the reduced load on client resolver and reduction in client latency is a big motivation to move this logic to the server-side. 
+The search path currently includes: 
+
+1. "$NS.svc.$SUFFIX"
+2. "svc.$SUFFIX"
+3. "$SUFFIX"
+4. Host level suffixes, which might be 2 or 3 in number.
+
+Where $NS stands for the namespace that the pod belongs to, $SUFFIX is the Kubernetes cluster suffix.
+
+These search paths are set to make sure: 
+
+ 1. Pods can discover Services in the same namespace using just the service name.
+ 2. Pods can discover Services across namespaces using shorthand of the form "$SVCNAME.$NSNAME"
+ 3. Pods can discover resources within the same cluster.
+
+These searchpaths are included in pods' /etc/resolv.conf by kubelet and are enforced by setting ndots to 5. This means any hostname lookups with fewer than 5 dots will be expanded using all the search paths listed.
+
+When pod issues a query to lookup hostname "service123", it is expanded to 6 queries - one for the original hostname and one with each of the searchpaths appended. Some resolvers issue both A and AAAA queries, so this can be a total of 12 or more queries for every single DNS lookup. When these queries are issued in parallel, they end up at the node with the same source tuple and need to be DNAT'ed increasing the chance of a [netfilter race condition](https://www.weave.works/blog/racy-conntrack-and-dns-lookup-timeouts).
+If one of the several queries fails, the DNS lookup on the client side will fail after a 5s timeout. This issue is most-applicable to A/AAAA lookups which form the bulk of DNS lookups. PTR Records are not subject to searchpath expansion(since they always have >5 dots).
```

Comment:
```suggestion
If a response is not received for one of the queries, the DNS lookup on the client side will fail after a 5s timeout. This issue is most-applicable to A/AAAA lookups which form the bulk of DNS lookups. PTR Records are not subject to searchpath expansion(since they always have >5 dots).
```